sslr-core/src/main/java/com/sonar/sslr/api/typed/AstNodeBuilder.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * SonarSource Language Recognizer
+ * Copyright (C) 2010 SonarSource
+ * dev@sonar.codehaus.org
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ */
+package com.sonar.sslr.api.typed;
+
+import com.google.common.annotations.Beta;
+import com.sonar.sslr.api.AstNode;
+import com.sonar.sslr.api.Rule;
+import com.sonar.sslr.api.Token;
+import com.sonar.sslr.api.TokenType;
+import com.sonar.sslr.api.Trivia;
+import com.sonar.sslr.impl.typed.SyntaxTreeCreator;
+import org.sonar.sslr.grammar.GrammarRuleKey;
+
+import java.util.List;
+
+@Beta
+public class AstNodeBuilder implements NodeBuilder {
+
+  @Override
+  public AstNode createNonTerminal(GrammarRuleKey ruleKey, Rule rule, List<Object> children, int startIndex, int endIndex) {
+    Token token = null;
+
+    for (Object child : children) {
+      if (child instanceof AstNode && ((AstNode) child).hasToken()) {
+        token = ((AstNode) child).getToken();
+        break;
+      }
+    }
+    AstNode astNode = new AstNode(rule, ruleKey.toString(), token);
+    for (Object child : children) {
+      astNode.addChild((AstNode) child);
+    }
+
+    astNode.setFromIndex(startIndex);
+    astNode.setToIndex(endIndex);
+
+    return astNode;
+  }
+
+  @Override
+  public AstNode createTerminal(Input input, int startIndex, int endIndex, List<Trivia> trivias, TokenType type) {
+    int[] lineAndColumn = input.lineAndColumnAt(startIndex);
+    Token token = Token.builder()
+      .setType(type == null ? UNDEFINED_TOKEN_TYPE : type)
+      .setLine(lineAndColumn[0])
+      .setColumn(lineAndColumn[1] - 1)
+      .setValueAndOriginalValue(input.substring(startIndex, endIndex))
+      .setURI(input.uri())
+      .setGeneratedCode(false)
+      .setTrivia(trivias)
+      .build();
+    AstNode astNode = new AstNode(token);
+    astNode.setFromIndex(startIndex);
+    astNode.setToIndex(endIndex);
+    return astNode;
+  }
+
+  private static final TokenType UNDEFINED_TOKEN_TYPE = new TokenType() {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Reduce this anonymous class number of lines from 23 to at most 20, or make it a named class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1188)